CommonTools/PileupAlgos/plugins/PuppiProducer.cc

```diff
@@ -100,40 +111,72 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
         if(iV->trackWeight(pPF->trackRef())>0) {
             closestVtx  = &(*iV);
             break;
-          }
+          }        
+        // in case it's unassocciated, keep more info
+        double tmpdz = 99999;
+        if      ( pPF->trackRef().isNonnull()    ) tmpdz = pPF->trackRef()   ->dz(iV->position());
+        else if ( pPF->gsfTrackRef().isNonnull() ) tmpdz = pPF->gsfTrackRef()->dz(iV->position());
+        if (fabs(tmpdz) < curdz){
+          curdz = fabs(tmpdz);
+          closestVtxForUnassociateds = pVtxId;
+        }
         pVtxId++;
+
       }
-    } else if(lPack->vertexRef().isNonnull() )  {
-        pDZ        = lPack->dz();
-        pD0        = lPack->dxy();
-        closestVtx = &(*(lPack->vertexRef()));
-        pVtxId = (lPack->fromPV() != (pat::PackedCandidate::PVUsedInFit));
-        if( (lPack->fromPV() == pat::PackedCandidate::PVLoose) || (lPack->fromPV() == pat::PackedCandidate::PVTight) ){
-            pVtxId = 0;
-        } 
-    }
+      int tmpFromPV = 0;  
+      // mocking the miniAOD definitions
+      if (closestVtx != 0 && fabs(pReco.charge) > 0 && pVtxId > 0) tmpFromPV = 0;
```

Comment:
these if blocks appear pretty wasteful and are probably less readable than nested ones.
Especially when the same condition repeats in every single one.